language/stdlib/modules/Roles.move

```diff
@@ -0,0 +1,265 @@
+address 0x1 {
+/// This module describes two things:
+/// 1. The relationship between roles, e.g. Role_A can creates accounts of Role_B
+/// 2. The granting of privileges to an account with a specific role
+/// It is important to note here that this module _does not_ describe the
+/// privileges that a specific role can have. This is a property of each of
+/// the modules that declares a privilege.
+///
+/// It also defines a capability for extracting capabilities from an
+/// account, and ensuring that they can only be "restored" back to the
+/// account that they were extracted from.
+///
+/// Roles are defined to be completely opaque outside of this module --
+/// all operations should be guarded by privilege checks, and not by role
+/// checks.
+module Roles {
+    use 0x1::Signer;
+    use 0x1::CoreAddresses;
+
+    /// TODO: This needs to be removed.
+    public fun parent_vasp_role_TESTNET_HACK(): ParentVASPRole {
+        ParentVASPRole{}
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Role ID constants
+    ///////////////////////////////////////////////////////////////////////////
+
+    // TODO: Change these to constants once the source language has them
+    public fun ASSOCIATION_ROOT_ROLE_ID(): u64 { 0 }
+    public fun TREASURY_COMPLIANCE_ROLE_ID(): u64 { 1 }
+    public fun DESIGNATED_DEALER_ROLE_ID(): u64 { 2 }
+    public fun VALIDATOR_ROLE_ID(): u64 { 3 }
+    public fun VALIDATOR_OPERATOR_ROLE_ID(): u64 { 4 }
+    public fun PARENT_VASP_ROLE_ID(): u64 { 5 }
+    public fun CHILD_VASP_ROLE_ID(): u64 { 6 }
+    public fun UNHOSTED_ROLE_ID(): u64 { 7 }
+
+    /// The roleId contains the role id for the account. This is only moved
+    /// to an account as a top-level resource, and is otherwise immovable.
+    /// INVARIANT: Once an account at address `A` is granted a role `R` it
+    ///            will remain an account with role `R` for all time.
+    resource struct RoleId {
+        role_id: u64,
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Privileges & Capabilities
+    ///////////////////////////////////////////////////////////////////////////
+
+    /// Privileges are extracted in to capabilities. Capabilities hold /
+    /// the account address that they were extracted from (i.e. tagged or
+    /// "tainted"). Capabilities can then only be restored to the account
+    /// where they were extracted from.
+    resource struct Capability<Privilege: resource> {
+        owner_address: address,
+    }
+
+    /// The internal representation of of a privilege. We wrap every
+    /// privilege witness resource here to avoid having to write extractors/restorers
+    /// for each privilege, but can instead write this generically.
+    resource struct Privilege<Priv: resource>  {
```

Comment:
What do you think about adding an `is_extracted` bool to guarantee that there is at most one `Capability<T>` for a `Privilege<T>` under a particular account at a time?